lib/elixir/lib/inspect.ex

```diff
@@ -86,17 +84,58 @@ defimpl Inspect, for: Atom do
           "Elixir." <> rest = binary
           rest
         end
-      valid_atom_identifier?(binary) ->
-        ":" <> binary
-      atom in [:%{}, :{}, :<<>>, :..., :%] ->
-        ":" <> binary
-      atom in Macro.binary_ops or atom in Macro.unary_ops ->
+      type when type in [:callable, :non_callable] ->
         ":" <> binary
+      :atom ->
+        ":" <> escape(binary)
+    end
+  end
+
+  # Classifies the given atom into one of five categories:
+  #
+  #   * :boolean_or_nil - the atoms true, false, nil
+  #
+  #   * :alias - a valid Elixir alias, like Foo, Foo.Bar and so on
+  #
+  #   * :callable - an atom that can be used as a function call after the
+  #     . operator (for example, :<> is callable because Foo.<>(1, 2, 3) is valid
+  #     syntax); this category includes identifiers like :foo
+  #
+  #   * :non_callable - an atom that cannot be used as a function call after the
+  #     . operator (for example, :<<>> is not callable because Foo.<<>> is a
+  #     syntax error); this category includes atoms like :Foo, since they are
+  #     valid identifiers but they need quotes to be used in function calls
+  #     (Foo."Bar")
+  #
+  #   * :atom - any other atom (these are usually escaped when inspected, like
+  #     :"foo and bar")
+  def classify(atom) do
+    binary = Atom.to_string(atom)
+
+    cond do
+      is_boolean(atom) or is_nil(atom) ->
+        :boolean_or_nil
+      valid_alias?(binary) ->
+        :alias
+      valid_atom_identifier?(binary) ->
+        if :binary.first(binary) in ?A..?Z do
+          :non_callable
+        else
+          :callable
+        end
+      atom in [:%{}, :{}, :<<>>, :..., :.., :.] ->
```

Comment:
I believe `{}` is callable but it uses a special syntax: `Foo.{1, 2, 3}`. I don't know what we want to make out of it.